File: Platform/infra/security/tests/ts/integration/auth-integration.test.ts
import { describe, it, expect, beforeAll, afterAll, beforeEach, afterEach } from '@jest/globals';
import { AuthClient, AuthClientFactory } from '../../../typescript/auth-client';
import {
    AuthConfig,
    AuthTokenResponse,
    AuthServerDiscoveryResponse,
    AuthTokenStorage,
    AuthLogger
} from '../../../typescript/auth-client';

/**
 * Integration tests for AuthClient against a real authentication server
 * 
 * These tests require a running authentication server and should be run in a test environment.
 * The tests can be configured to run against different authentication providers by setting
 * environment variables.
 * 
 * Environment Variables:
 * - AUTH_TEST_SERVER_URL: Base URL of the authentication server
 * - AUTH_TEST_CLIENT_ID: Client ID for testing
 * - AUTH_TEST_CLIENT_SECRET: Client secret for testing
 * - AUTH_TEST_REDIRECT_URI: Redirect URI for testing
 * - AUTH_TEST_USERNAME: Username for resource owner password credentials
 * - AUTH_TEST_PASSWORD: Password for resource owner password credentials
 * - AUTH_SKIP_INTEGRATION_TESTS: Set to 'true' to skip integration tests
 */

// Check if integration tests should be skipped
const SKIP_INTEGRATION_TESTS = process.env.AUTH_SKIP_INTEGRATION_TESTS === 'true';

// Test server configuration
const TEST_CONFIG = {
    serverUrl: process.env.AUTH_TEST_SERVER_URL || 'https://localhost:5001',
    clientId: process.env.AUTH_TEST_CLIENT_ID || 'test-client-id',
    clientSecret: process.env.AUTH_TEST_CLIENT_SECRET || 'test-client-secret',
    redirectUri: process.env.AUTH_TEST_REDIRECT_URI || 'https://localhost:3000/callback',
    username: process.env.AUTH_TEST_USERNAME || 'testuser',
    password: process.env.AUTH_TEST_PASSWORD || 'testpass'
};

// Simple in-memory token storage for integration tests
class IntegrationTokenStorage implements AuthTokenStorage {
    private tokens: Map<string, string> = new Map();

    async getToken(key: string): Promise<string | null> {
        return this.tokens.get(key) || null;
    }

    async setToken(key: string, token: string): Promise<void> {
        this.tokens.set(key, token);
    }

    async removeToken(key: string): Promise<void> {
        this.tokens.delete(key);
    }

    async clear(): Promise<void> {
        this.tokens.clear();
    }
}

// Simple console logger for integration tests
class IntegrationLogger implements AuthLogger {
    debug(message: string, data?: any): void {
        if (process.env.AUTH_TEST_DEBUG === 'true') {
            console.debug(`[DEBUG] ${message}`, data || '');
        }
    }

    info(message: string, data?: any): void {
        console.info(`[INFO] ${message}`, data || '');
    }

    warn(message: string, data?: any): void {
        console.warn(`[WARN] ${message}`, data || '');
    }

    error(message: string, data?: any): void {
        console.error(`[ERROR] ${message}`, data || '');
    }
}

// Test helper to wait for server availability
async function waitForServerAvailability(url: string, timeoutMs: number = 30000): Promise<boolean> {
    const startTime = Date.now();
    while (Date.now() - startTime < timeoutMs) {
        try {
            const response = await fetch(url);
            if (response.status < 500) {
                return true;
            }
        } catch (error) {
            // Server not available yet
        }
        await new Promise(resolve => setTimeout(resolve, 1000));
    }
    return false;
}

// Skip integration tests if configured
const describeIntegration = SKIP_INTEGRATION_TESTS ? describe.skip : describe;

describeIntegration('AuthClient Integration Tests', () => {
    let client: AuthClient;
    let config: AuthConfig;
    let tokenStorage: IntegrationTokenStorage;
    let logger: IntegrationLogger;

    beforeAll(async () => {
        // Check if test server is available
        const serverAvailable = await waitForServerAvailability(`${TEST_CONFIG.serverUrl}/.well-known/openid_configuration`);
        if (!serverAvailable) {
            console.warn(`Authentication test server not available at ${TEST_CONFIG.serverUrl}. Skipping integration tests.`);
            return;
        }

        // Initialize test components
        tokenStorage = new IntegrationTokenStorage();
        logger = new IntegrationLogger();

        config = {
            clientId: TEST_CONFIG.clientId,
            clientSecret: TEST_CONFIG.clientSecret,
            authorizationUrl: `${TEST_CONFIG.serverUrl}/oauth2/authorize`,
            tokenUrl: `${TEST_CONFIG.serverUrl}/oauth2/token`,
            introspectionUrl: `${TEST_CONFIG.serverUrl}/oauth2/introspect`,
            revocationUrl: `${TEST_CONFIG.serverUrl}/oauth2/revoke`,
            discoveryUrl: `${TEST_CONFIG.serverUrl}/.well-known/openid_configuration`,
            redirectUri: TEST_CONFIG.redirectUri,
            scopes: ['read', 'write'],
            enableAutoRefresh: true,
            tokenRefreshThresholdSeconds: 300,
            maxRetryAttempts: 3,
            retryDelayMs: 1000,
            requestTimeoutMs: 30000,
            enableServerDiscovery: true
        };

        client = new AuthClient(config, tokenStorage, logger);
    }, 60000);

    beforeEach(async () => {
        // Clear stored tokens before each test
        await tokenStorage.clear();
    });

    afterEach(async () => {
        // Clean up any stored tokens after each test
        await tokenStorage.clear();
    });

    describe('Server Discovery', () => {
        it('should successfully discover server endpoints', async () => {
            const discovery = await client.discoverServerEndpoints();

            expect(discovery).toBeDefined();
            expect(discovery.issuer).toBeTruthy();
            expect(discovery.authorization_endpoint).toBeTruthy();
            expect(discovery.token_endpoint).toBeTruthy();
            expect(discovery.grant_types_supported).toContain('client_credentials');
            expect(discovery.response_types_supported).toContain('code');
        });

        it('should handle server discovery failures gracefully', async () => {            const invalidClient = new AuthClient(
                { ...config, discoveryUrl: 'https://invalid.example.com/.well-known/openid_configuration' },
                tokenStorage,
                logger
            );

            await expect(invalidClient.discoverServerEndpoints())
                .rejects.toThrow();
        });
    });

    describe('Client Credentials Flow', () => {
        it('should successfully obtain access token using client credentials', async () => {
            const tokenResponse = await client.getTokenWithClientCredentials();

            expect(tokenResponse).toBeDefined();
            expect(tokenResponse.access_token).toBeTruthy();
            expect(tokenResponse.token_type).toBe('Bearer');
            expect(tokenResponse.expires_in).toBeGreaterThan(0);
            expect(typeof tokenResponse.expires_in).toBe('number');

            // Verify token is stored
            const storedToken = await tokenStorage.getToken('access_token');
            expect(storedToken).toBe(tokenResponse.access_token);
        });

        it('should successfully obtain token with specific scopes', async () => {
            const tokenResponse = await client.getTokenWithClientCredentials(['read']);

            expect(tokenResponse).toBeDefined();
            expect(tokenResponse.access_token).toBeTruthy();
            expect(tokenResponse.scope).toContain('read');
        });

        it('should handle invalid client credentials', async () => {
            const invalidClient = new AuthClient(
                { ...config, clientSecret: 'invalid-secret' },
                tokenStorage,
                logger
            );

            await expect(invalidClient.getTokenWithClientCredentials())
                .rejects.toThrow();
        });
    });

    describe('Token Introspection', () => {
        let accessToken: string;

        beforeEach(async () => {
            // Obtain a valid access token for introspection tests
            const tokenResponse = await client.getTokenWithClientCredentials();
            accessToken = tokenResponse.access_token;
        });

        it('should successfully introspect active token', async () => {
            const introspection = await client.introspectToken(accessToken);

            expect(introspection).toBeDefined();
            expect(introspection.active).toBe(true);
            expect(introspection.client_id).toBe(TEST_CONFIG.clientId);
            expect(introspection.token_type).toBe('Bearer');
            expect(introspection.exp).toBeGreaterThan(Math.floor(Date.now() / 1000));
        });

        it('should return inactive for invalid token', async () => {
            const introspection = await client.introspectToken('invalid-token');

            expect(introspection).toBeDefined();
            expect(introspection.active).toBe(false);
        });
    });

    describe('Token Revocation', () => {
        let accessToken: string;

        beforeEach(async () => {
            // Obtain a valid access token for revocation tests
            const tokenResponse = await client.getTokenWithClientCredentials();
            accessToken = tokenResponse.access_token;
        });

        it('should successfully revoke access token', async () => {
            await expect(client.revokeToken(accessToken, 'access_token'))
                .resolves.not.toThrow();

            // Verify token is now inactive
            const introspection = await client.introspectToken(accessToken);
            expect(introspection.active).toBe(false);
        });

        it('should handle revocation of already revoked token', async () => {
            // Revoke the token first
            await client.revokeToken(accessToken, 'access_token');

            // Try to revoke again - should not throw
            await expect(client.revokeToken(accessToken, 'access_token'))
                .resolves.not.toThrow();
        });
    });

    describe('Token Storage Integration', () => {
        it('should store and retrieve tokens correctly', async () => {
            const tokenResponse = await client.getTokenWithClientCredentials();

            // Verify tokens are stored
            const storedAccessToken = await client.getStoredAccessToken();
            expect(storedAccessToken).toBe(tokenResponse.access_token);

            if (tokenResponse.refresh_token) {
                const storedRefreshToken = await client.getStoredRefreshToken();
                expect(storedRefreshToken).toBe(tokenResponse.refresh_token);
            }
        });

        it('should clear stored tokens', async () => {
            // First obtain and store tokens
            await client.getTokenWithClientCredentials();

            // Verify tokens are stored
            expect(await client.getStoredAccessToken()).toBeTruthy();

            // Clear tokens
            await client.clearStoredTokens();

            // Verify tokens are cleared
            expect(await client.getStoredAccessToken()).toBeNull();
            expect(await client.getStoredRefreshToken()).toBeNull();
        });
    });

    describe('Auto-Refresh Functionality', () => {
        it('should detect when token needs refresh', async () => {
            // Get a fresh token
            const tokenResponse = await client.getTokenWithClientCredentials();
            
            // Token should not need refresh initially
            const needsRefresh = await client.needsTokenRefresh();
            expect(needsRefresh).toBe(false);
        });

        it('should provide valid access token (with auto-refresh if needed)', async () => {
            const accessToken = await client.getValidAccessToken();
            
            expect(accessToken).toBeTruthy();
            
            // Verify the token is active
            const introspection = await client.introspectToken(accessToken);
            expect(introspection.active).toBe(true);
        });
    });

    describe('Error Handling and Resilience', () => {
        it('should handle network timeouts gracefully', async () => {
            const shortTimeoutClient = new AuthClient(
                { ...config, requestTimeoutMs: 1 }, // Very short timeout
                tokenStorage,
                logger
            );

            await expect(shortTimeoutClient.getTokenWithClientCredentials())
                .rejects.toThrow();
        });

        it('should retry failed requests', async () => {
            // This test would require a way to simulate intermittent failures
            // For now, we just verify that the retry configuration is honored
            const retryClient = new AuthClient(
                { ...config, maxRetryAttempts: 2, retryDelayMs: 100 },
                tokenStorage,
                logger
            );

            // Should still succeed with retries enabled
            const tokenResponse = await retryClient.getTokenWithClientCredentials();
            expect(tokenResponse.access_token).toBeTruthy();
        });
    });

    describe('Authorization Code Flow (Manual)', () => {
        it('should generate correct authorization URL', () => {
            const state = 'test-state-' + Date.now();
            const scopes = ['read', 'write'];
            
            const authUrl = client.getAuthorizationUrl(scopes, state);
            
            const url = new URL(authUrl);
            expect(url.origin + url.pathname).toBe(config.authorizationUrl);
            expect(url.searchParams.get('client_id')).toBe(config.clientId);
            expect(url.searchParams.get('response_type')).toBe('code');
            expect(url.searchParams.get('redirect_uri')).toBe(config.redirectUri);
            expect(url.searchParams.get('scope')).toBe('read write');
            expect(url.searchParams.get('state')).toBe(state);
        });

        it('should generate authorization URL with PKCE', () => {
            const state = 'test-state-' + Date.now();
            const scopes = ['read'];
            const pkceParams = {
                code_verifier: 'test-code-verifier',
                code_challenge: 'test-code-challenge',
                code_challenge_method: 'S256'
            };
            
            const authUrl = client.getAuthorizationUrl(scopes, state, pkceParams);
            
            const url = new URL(authUrl);
            expect(url.searchParams.get('code_challenge')).toBe(pkceParams.code_challenge);
            expect(url.searchParams.get('code_challenge_method')).toBe(pkceParams.code_challenge_method);
        });

        // Note: Testing the full authorization code flow requires manual intervention
        // or a test harness that can simulate browser interactions
        it.skip('should exchange authorization code for token', async () => {
            // This test would require a valid authorization code obtained through browser flow
            const authCode = 'test-authorization-code';
            
            const tokenResponse = await client.getTokenWithAuthorizationCode(authCode);
            
            expect(tokenResponse.access_token).toBeTruthy();
            expect(tokenResponse.token_type).toBe('Bearer');
        });
    });

    describe('Concurrent Operations', () => {
        it('should handle concurrent token requests', async () => {
            // Make multiple concurrent requests
            const promises = Array.from({ length: 3 }, () => 
                client.getTokenWithClientCredentials()
            );

            const results = await Promise.all(promises);

            // All requests should succeed
            results.forEach(result => {
                expect(result.access_token).toBeTruthy();
                expect(result.token_type).toBe('Bearer');
            });
        });

        it('should handle concurrent storage operations', async () => {
            // Make concurrent storage operations
            const setPromises = Array.from({ length: 5 }, (_, i) => 
                tokenStorage.setToken(`test_token_${i}`, `value_${i}`)
            );

            await Promise.all(setPromises);

            // Verify all tokens were stored
            const getPromises = Array.from({ length: 5 }, (_, i) => 
                tokenStorage.getToken(`test_token_${i}`)
            );

            const values = await Promise.all(getPromises);
            values.forEach((value, i) => {
                expect(value).toBe(`value_${i}`);
            });
        });
    });
});

describeIntegration('AuthClientFactory Integration Tests', () => {
    let tokenStorage: IntegrationTokenStorage;
    let logger: IntegrationLogger;

    beforeAll(() => {
        tokenStorage = new IntegrationTokenStorage();
        logger = new IntegrationLogger();
    });

    it('should create client from environment variables', () => {
        // Mock environment variables
        const originalEnv = process.env;
        process.env = {
            ...originalEnv,            AUTH_CLIENT_ID: TEST_CONFIG.clientId,
            AUTH_CLIENT_SECRET: TEST_CONFIG.clientSecret,
            AUTH_TOKEN_URL: `${TEST_CONFIG.serverUrl}/oauth2/token`,
            AUTH_AUTHORIZATION_URL: `${TEST_CONFIG.serverUrl}/oauth2/authorize`
        };

        try {            const client = AuthClientFactory.createFromEnvironment(tokenStorage, logger);
            expect(client).toBeInstanceOf(AuthClient);
        } finally {
            process.env = originalEnv;
        }
    });

    it('should create client with discovery URL only', async () => {
        const discoveryConfig = {
            clientId: TEST_CONFIG.clientId,
            clientSecret: TEST_CONFIG.clientSecret,            discoveryUrl: `${TEST_CONFIG.serverUrl}/.well-known/openid_configuration`
        };

        const client = AuthClientFactory.create(discoveryConfig, tokenStorage, logger);
        expect(client).toBeInstanceOf(AuthClient);

        // Verify it can discover and use endpoints
        const discovery = await client.discoverServerEndpoints();
        expect(discovery.token_endpoint).toBeTruthy();
    });
});

// Performance and load testing
describeIntegration('AuthClient Performance Tests', () => {
    let client: OAuth2AuthClient;
    let tokenStorage: IntegrationTokenStorage;
    let logger: IntegrationLogger;

    beforeAll(() => {
        tokenStorage = new IntegrationTokenStorage();
        logger = new IntegrationLogger();

        const config: OAuth2Config = {
            clientId: TEST_CONFIG.clientId,
            clientSecret: TEST_CONFIG.clientSecret,
            tokenUrl: `${TEST_CONFIG.serverUrl}/oauth2/token`,
            maxRetryAttempts: 1, // Reduce retries for performance tests
            requestTimeoutMs: 10000
        };

        client = new OAuth2AuthClient(config, tokenStorage, logger);
    });

    it('should handle high-frequency token requests efficiently', async () => {
        const startTime = Date.now();
        const requestCount = 10;

        // Make multiple sequential requests
        const promises = Array.from({ length: requestCount }, () => 
            client.getTokenWithClientCredentials()
        );

        const results = await Promise.all(promises);
        const endTime = Date.now();

        // All requests should succeed
        expect(results).toHaveLength(requestCount);
        results.forEach(result => {
            expect(result.access_token).toBeTruthy();
        });

        // Performance assertion (should complete within reasonable time)
        const totalTime = endTime - startTime;
        const averageTime = totalTime / requestCount;
        console.info(`Average request time: ${averageTime}ms`);
        
        // Should average less than 5 seconds per request
        expect(averageTime).toBeLessThan(5000);
    }, 60000);
});

File: Platform/infra/security/tests/cpp/integration/auth_integration_test.cpp
#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <thread>
#include <future>
#include <chrono>
#include <cstdlib>
#include <memory>

#include "auth_client.h"
#include "oauth2_test_mocks.h"

namespace coyote_sense {
namespace oauth2 {
namespace test {

/**
 * Integration tests for OAuth2AuthClient against real OAuth2 server
 */
class OAuth2IntegrationTest : public ::testing::Test {
protected:
    void SetUp() override {
        // Load configuration from environment variables
        config_ = std::make_unique<OAuth2ClientConfiguration>();
        config_->server_url = GetEnvVar("OAUTH2_SERVER_URL", "https://localhost:5001");
        config_->client_id = GetEnvVar("OAUTH2_CLIENT_ID", "integration-test-client");
        config_->client_secret = GetEnvVar("OAUTH2_CLIENT_SECRET", "integration-test-secret");
        config_->scope = GetEnvVar("OAUTH2_SCOPE", "api.read api.write");
        config_->enable_auto_refresh = true;
        config_->retry_policy.max_retries = 3;
        config_->retry_policy.base_delay = std::chrono::seconds(1);
        config_->retry_policy.max_delay = std::chrono::seconds(10);
        config_->retry_policy.use_exponential_backoff = true;

        // Create test implementations
        token_storage_ = std::make_shared<TestOAuth2TokenStorage>();
        logger_ = std::make_shared<TestOAuth2Logger>();
        
        // Create client
        client_ = std::make_unique<OAuth2AuthClient>(*config_, token_storage_, logger_);
        
        // Check if server is available
        server_available_ = IsServerAvailable();
    }

    void TearDown() override {
        client_.reset();
    }

    std::string GetEnvVar(const std::string& name, const std::string& default_value) {
        const char* value = std::getenv(name.c_str());
        return value ? std::string(value) : default_value;
    }

    bool IsServerAvailable() {
        try {
            auto http_client = std::make_shared<HttpClient>();
            auto response = http_client->Get(config_->server_url + "/.well-known/openid_configuration");
            return response.status_code == 200;
        } catch (...) {
            return false;
        }
    }

    std::unique_ptr<OAuth2ClientConfiguration> config_;
    std::shared_ptr<TestOAuth2TokenStorage> token_storage_;
    std::shared_ptr<TestOAuth2Logger> logger_;
    std::unique_ptr<OAuth2AuthClient> client_;
    bool server_available_ = false;
};

TEST_F(OAuth2IntegrationTest, ClientCredentialsFlow_ShouldAuthenticateSuccessfully) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Act
    auto future = client_->AuthenticateClientCredentialsAsync();
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_FALSE(result->access_token.empty());
    EXPECT_EQ(result->token_type, "Bearer");
    EXPECT_GT(result->expires_in, 0);
}

TEST_F(OAuth2IntegrationTest, JwtBearerFlow_WithValidJwt_ShouldAuthenticateSuccessfully) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Get a valid token to use as JWT
    auto client_creds_future = client_->AuthenticateClientCredentialsAsync();
    auto client_creds_result = client_creds_future.get();
    ASSERT_TRUE(client_creds_result.has_value());
    std::string jwt_token = client_creds_result->access_token;

    // Act
    auto future = client_->AuthenticateJwtBearerAsync(jwt_token);
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_FALSE(result->access_token.empty());
    EXPECT_EQ(result->token_type, "Bearer");
}

TEST_F(OAuth2IntegrationTest, TokenIntrospection_WithValidToken_ShouldReturnActiveToken) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Get a valid token
    auto auth_future = client_->AuthenticateClientCredentialsAsync();
    auto auth_result = auth_future.get();
    ASSERT_TRUE(auth_result.has_value());
    std::string access_token = auth_result->access_token;

    // Act
    auto future = client_->IntrospectTokenAsync(access_token);
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_TRUE(result->active);
    EXPECT_EQ(result->client_id, config_->client_id);
}

TEST_F(OAuth2IntegrationTest, TokenIntrospection_WithInvalidToken_ShouldReturnInactiveToken) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Act
    auto future = client_->IntrospectTokenAsync("invalid-token");
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_FALSE(result->active);
}

TEST_F(OAuth2IntegrationTest, TokenRevocation_WithValidToken_ShouldRevokeSuccessfully) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Get a valid token
    auto auth_future = client_->AuthenticateClientCredentialsAsync();
    auto auth_result = auth_future.get();
    ASSERT_TRUE(auth_result.has_value());
    std::string access_token = auth_result->access_token;

    // Act
    auto revoke_future = client_->RevokeTokenAsync(access_token);
    bool revoke_result = revoke_future.get();

    // Assert
    EXPECT_TRUE(revoke_result);

    // Verify token is no longer active
    auto introspect_future = client_->IntrospectTokenAsync(access_token);
    auto introspect_result = introspect_future.get();
    ASSERT_TRUE(introspect_result.has_value());
    EXPECT_FALSE(introspect_result->active);
}

TEST_F(OAuth2IntegrationTest, ServerDiscovery_ShouldReturnValidEndpoints) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Act
    auto future = client_->DiscoverServerEndpointsAsync();
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_FALSE(result->token_endpoint.empty());
    EXPECT_FALSE(result->introspection_endpoint.empty());
    EXPECT_FALSE(result->revocation_endpoint.empty());
    EXPECT_FALSE(result->supported_grant_types.empty());
    
    // Check for client_credentials support
    auto& grant_types = result->supported_grant_types;
    bool supports_client_credentials = std::find(grant_types.begin(), grant_types.end(), 
                                                "client_credentials") != grant_types.end();
    EXPECT_TRUE(supports_client_credentials);
}

TEST_F(OAuth2IntegrationTest, ConcurrentAuthentication_ShouldHandleMultipleRequests) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange
    const int num_concurrent_requests = 5;
    std::vector<std::future<std::optional<OAuth2TokenResponse>>> futures;

    // Act - Create multiple concurrent authentication requests
    for (int i = 0; i < num_concurrent_requests; ++i) {
        futures.push_back(client_->AuthenticateClientCredentialsAsync());
    }

    // Wait for all requests to complete
    std::vector<std::optional<OAuth2TokenResponse>> results;
    for (auto& future : futures) {
        results.push_back(future.get());
    }

    // Assert
    EXPECT_EQ(results.size(), num_concurrent_requests);
    for (const auto& result : results) {
        ASSERT_TRUE(result.has_value());
        EXPECT_FALSE(result->access_token.empty());
        EXPECT_EQ(result->token_type, "Bearer");
    }
}

TEST_F(OAuth2IntegrationTest, AutoRefresh_WhenTokenExpires_ShouldRefreshAutomatically) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Get initial token
    auto initial_future = client_->AuthenticateClientCredentialsAsync();
    auto initial_result = initial_future.get();
    ASSERT_TRUE(initial_result.has_value());
    std::string initial_token = initial_result->access_token;

    // Wait for potential token expiry simulation
    std::this_thread::sleep_for(std::chrono::seconds(2));

    // Act - Request new token (should trigger auto-refresh if needed)
    auto refreshed_future = client_->AuthenticateClientCredentialsAsync();
    auto refreshed_result = refreshed_future.get();

    // Assert
    ASSERT_TRUE(refreshed_result.has_value());
    EXPECT_FALSE(refreshed_result->access_token.empty());
    EXPECT_EQ(refreshed_result->token_type, "Bearer");
    // Token might be the same if still valid, or different if refreshed
}

TEST_F(OAuth2IntegrationTest, InvalidCredentials_ShouldReturnFailure) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Create client with invalid credentials
    OAuth2ClientConfiguration invalid_config = *config_;
    invalid_config.client_id = "invalid-client-id";
    invalid_config.client_secret = "invalid-client-secret";

    auto invalid_token_storage = std::make_shared<TestOAuth2TokenStorage>();
    auto invalid_logger = std::make_shared<TestOAuth2Logger>();
    auto invalid_client = std::make_unique<OAuth2AuthClient>(
        invalid_config, invalid_token_storage, invalid_logger);

    // Act
    auto future = invalid_client->AuthenticateClientCredentialsAsync();
    auto result = future.get();

    // Assert
    EXPECT_FALSE(result.has_value());
    
    // Check that error was logged
    auto& log_messages = invalid_logger->GetLogMessages();
    bool error_logged = false;
    for (const auto& message : log_messages) {
        if (message.find("error") != std::string::npos || 
            message.find("unauthorized") != std::string::npos ||
            message.find("invalid") != std::string::npos) {
            error_logged = true;
            break;
        }
    }
    EXPECT_TRUE(error_logged);
}

TEST_F(OAuth2IntegrationTest, HealthCheck_ShouldReturnServerStatus) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Act
    auto future = client_->CheckServerHealthAsync();
    bool health_status = future.get();

    // Assert
    EXPECT_TRUE(health_status);
}

TEST_F(OAuth2IntegrationTest, LargeScope_ShouldHandleExtensivePermissions) {
    // Skip if OAuth2 server is not available
    if (!server_available_) {
        GTEST_SKIP() << "OAuth2 server is not available, skipping integration test";
    }

    // Arrange - Create client with large scope
    OAuth2ClientConfiguration large_scope_config = *config_;
    large_scope_config.scope = "api.read api.write api.admin openid profile email";

    auto large_scope_storage = std::make_shared<TestOAuth2TokenStorage>();
    auto large_scope_logger = std::make_shared<TestOAuth2Logger>();
    auto large_scope_client = std::make_unique<OAuth2AuthClient>(
        large_scope_config, large_scope_storage, large_scope_logger);

    // Act
    auto future = large_scope_client->AuthenticateClientCredentialsAsync();
    auto result = future.get();

    // Assert
    ASSERT_TRUE(result.has_value());
    EXPECT_FALSE(result->access_token.empty());
    // The scope in response might be limited by what the server supports
}

TEST_F(OAuth2IntegrationTest, NetworkTimeout_ShouldHandleGracefully) {
    // Arrange - Create client with very short timeout
    OAuth2ClientConfiguration timeout_config = *config_;
    timeout_config.timeout = std::chrono::milliseconds(1); // Very short timeout
    timeout_config.server_url = "https://httpbin.org/delay/10"; // Slow endpoint

    auto timeout_storage = std::make_shared<TestOAuth2TokenStorage>();
    auto timeout_logger = std::make_shared<TestOAuth2Logger>();
    auto timeout_client = std::make_unique<OAuth2AuthClient>(
        timeout_config, timeout_storage, timeout_logger);

    // Act
    auto future = timeout_client->AuthenticateClientCredentialsAsync();
    auto result = future.get();

    // Assert
    EXPECT_FALSE(result.has_value());
    
    // Check that timeout error was logged
    auto& log_messages = timeout_logger->GetLogMessages();
    bool timeout_logged = false;
    for (const auto& message : log_messages) {
        if (message.find("timeout") != std::string::npos) {
            timeout_logged = true;
            break;
        }
    }
    EXPECT_TRUE(timeout_logged);
}

/**
 * Performance test implementation for C++ OAuth2 client
 */
class OAuth2PerformanceTest : public OAuth2IntegrationTest {
protected:
    void SetUp() override {
        OAuth2IntegrationTest::SetUp();
        
        // Only run performance tests in specific environments
        const char* run_perf = std::getenv("RUN_PERFORMANCE_TESTS");
        skip_performance_ = !(run_perf && std::string(run_perf) == "1");
    }

    bool skip_performance_ = false;
};

TEST_F(OAuth2PerformanceTest, HighConcurrency_ShouldMaintainPerformance) {
    if (skip_performance_ || !server_available_) {
        GTEST_SKIP() << "Performance tests disabled or server unavailable";
    }

    // Arrange
    const int num_concurrent_requests = 50;
    const int requests_per_thread = 10;
    std::vector<std::future<void>> futures;
    
    auto start_time = std::chrono::high_resolution_clock::now();

    // Act - Create concurrent authentication requests
    for (int i = 0; i < num_concurrent_requests; ++i) {
        futures.push_back(std::async(std::launch::async, [this, requests_per_thread]() {
            for (int j = 0; j < requests_per_thread; ++j) {
                auto future = client_->AuthenticateClientCredentialsAsync();
                auto result = future.get();
                EXPECT_TRUE(result.has_value());
            }
        }));
    }

    // Wait for all futures to complete
    for (auto& future : futures) {
        future.get();
    }
    
    auto end_time = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);

    // Assert performance metrics
    int total_requests = num_concurrent_requests * requests_per_thread;
    double requests_per_second = (total_requests * 1000.0) / duration.count();
    
    std::cout << "Performance Results:\n";
    std::cout << "Total Requests: " << total_requests << "\n";
    std::cout << "Duration: " << duration.count() << " ms\n";
    std::cout << "Requests/Second: " << requests_per_second << "\n";
    
    EXPECT_GT(requests_per_second, 5.0) << "Should handle at least 5 requests per second";
    EXPECT_LT(duration.count(), 30000) << "Should complete within 30 seconds";
}

TEST_F(OAuth2PerformanceTest, MemoryUsage_ShouldRemainStable) {
    if (skip_performance_ || !server_available_) {
        GTEST_SKIP() << "Performance tests disabled or server unavailable";
    }

    // This test would require platform-specific memory monitoring
    // For now, we'll just ensure no obvious memory leaks by running many operations
    
    const int iterations = 100;
    
    for (int i = 0; i < iterations; ++i) {
        auto auth_future = client_->AuthenticateClientCredentialsAsync();
        auto auth_result = auth_future.get();
        ASSERT_TRUE(auth_result.has_value());
        
        // Periodically perform introspection and revocation
        if (i % 10 == 0) {
            auto introspect_future = client_->IntrospectTokenAsync(auth_result->access_token);
            auto introspect_result = introspect_future.get();
            EXPECT_TRUE(introspect_result.has_value());
            
            auto revoke_future = client_->RevokeTokenAsync(auth_result->access_token);
            bool revoke_result = revoke_future.get();
            EXPECT_TRUE(revoke_result);
        }
    }
    
    // If we reach here without crashing, memory usage is likely stable
    SUCCEED();
}

} // namespace test
} // namespace oauth2
} // namespace coyote_sense
